storage/series/buffer.go

```diff
@@ -660,20 +688,78 @@ func (b *dbBufferBucket) discardMerged() discardMergedResult {
 		}
 	}
 
+	var lastWriteAt time.Time
 	iter := b.opts.MultiReaderIteratorPool().Get()
 	iter.Reset(readers)
+
+	defer func() {
+		iter.Close()
+		for _, stream := range readers {
+			if resource, ok := stream.(context.Finalizer); ok {
+				resource.Finalize()
+			}
+		}
+	}()
+
 	for iter.Next() {
 		dp, unit, annotation := iter.Current()
 		if err := encoder.Encode(dp, unit, annotation); err != nil {
-			log := b.opts.InstrumentOptions().Logger()
-			log.Errorf("buffer merge encode error: %v", err)
-			break
+			return mergeResult{}, err
 		}
+		lastWriteAt = dp.Timestamp
+	}
+	if err := iter.Err(); err != nil {
+		return mergeResult{}, err
+	}
+
+	b.resetEncoders()
+	b.resetBootstrapped()
+
+	b.encoders = append(b.encoders, inOrderEncoder{
+		lastWriteAt: lastWriteAt,
+		encoder:     encoder,
+	})
+
+	return mergeResult{merges: merges}, nil
+}
+
+type discardMergedResult struct {
+	block  block.DatabaseBlock
+	merges int
+}
+
+func (b *dbBufferBucket) discardMerged() (discardMergedResult, error) {
+	defer func() {
+		b.resetEncoders()
+		b.resetBootstrapped()
+		b.empty = true
+	}()
+
+	if b.hasJustSingleEncoder() {
+		// Already merged as a single encoder
+		encoder := b.encoders[0].encoder
+		newBlock := b.opts.DatabaseBlockOptions().DatabaseBlockPool().Get()
+		newBlock.Reset(b.start, encoder.Discard())
```

Comment:
Do we have to put the encoder back into the pool here?

Reply:
`encoder.Discard()` will recycle the encoder, it calls `.Close()` after stealing the buffer =]